### PR TITLE
An example on detecting keypoints

### PR DESCRIPTION
**Pull request by sayakpaul**
Keypoint detection is useful in many applications like image matching, pose classification, etc. 

This example shows how to use data augmentation and transfer learning to train a keypoint detector on the [SatnfordExtra](https://github.com/benjiebob/StanfordExtra) dataset. 

Since augmenting images can change the coordinates of the keypoints, I think this example would be a good reference for developers to know how to properly apply data augmentation in situations like this. On the other hand, transfer learning can easily be applied to this task for either feature extraction or fine-tuning.

### CODE REVIEW

File: examples/vision/keypoint_detection.py [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+
+There are a total of 24 entries present inside `joints`. Each entry has 3 values:
+
+* x-coordinate
+* y-coordinate
+* visibility flag of the keypoints (1 indicates visibility and 0 indicates non-visibility)
+
+As we can see `joints` contain multiple `[0, 0, 0]` entries which denote that those
+keypoints were not labeled. In this example, we will consider both the non-visible
+keypoints as and the unlabeled keypoints in order to allow mini-batch learning.
```
Comment (s-mrb) [inline examples/vision/keypoint_detection.py:159]: `as and`

File: examples/vision/keypoint_detection.py [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+
+    images.append(image)
+    keypoints.append(keypoint)
+
+visualize_keypoints(images, keypoints)
+
+"""
+The plots show that we have images of non-uniform sizes which is expected in most of the
+real-world scenarios. However, if we resize these images to have a uniform shape (for
+e.g. (224 x 224)) their grouth-truth annotations will also be affected for that. The same
```
Comment (s-mrb) [inline examples/vision/keypoint_detection.py:234]: grouth in `grouth-truth`

File: examples/vision/keypoint_detection.py [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+
+"""
+## Define augmentation transforms
+"""
+
+train_aug = iaa.Sequential(
+    [
+        iaa.Resize(IMG_SIZE, interpolation="linear"),
+        iaa.Fliplr(0.3),
+        iaa.Sometimes(0.3, iaa.Affine(rotate=10, scale=(0.5, 0.7))),
```
Comment (s-mrb) [inline examples/vision/keypoint_detection.py:326]: it would be great if you add a comment about what `Sometimes` doing.

**Review by s-mrb — COMMENTED** [review]
Hi @sayakpaul , checkout comments.

File: examples/vision/keypoint_detection.py [hunk 4]
```diff
@@ -0,0 +1,438 @@
+"""
+Title: Keypoint Detection with Transfer Learning
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/vision/keypoint_detection.py:2]: Apparently, that is not the case. You can refer to [here](https://keras.io/examples/vision/) to confirm.

File: examples/vision/keypoint_detection.py [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+Author: [Sayak Paul](https://twitter.com/RisingSayak)
+Date created: 2021/05/02
+Last modified: 2021/05/02
+Description: Training a keypoint detector with data augmentation and transfer learning.
+"""
+"""
+Keypoint detection helps to locate the key parts from objects. For example, the key parts
+in our face include nose tips, eyebrows, eye corners, and so on. These parts help to
+represent the underlying object in a feature-rich manner. Keypoint detection has
+applications ranging from pose estimation, face detection, etc.
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/vision/keypoint_detection.py:12]: Not sure what is meant here.

[the remark this answers is not in the document]
Comment (s-mrb) [reply examples/vision/keypoint_detection.py:12]: `ranging from` pose estimation `to` face detection. or something like this.

IG `ranging from` is accompanied by `to`, or else sentence feels incomplete.

[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/vision/keypoint_detection.py:12]: Got it. Changed the sentence wording. Thanks!

File: examples/vision/keypoint_detection.py [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+
+# Set up a dictionary, mapping all the ground-truth information
+# with respect to the path of the image.
+json_dict = {i["img_path"]: i for i in json_data}
+
+"""
+A single entry of `json_dict` looks like the following:
+
+```
+'n02085782-Japanese_spaniel/n02085782_2886.jpg': {'img_bbox': [205,
```
Comment (fchollet) [inline examples/vision/keypoint_detection.py:108]: Please pretty-print the JSON dict

File: examples/vision/keypoint_detection.py [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+            for idx, (x, y) in enumerate(current_keypoint):
+                ax_all.scatter([x], [y], c=colours[idx], marker="x", s=50, linewidths=5)
+
+    plt.tight_layout(pad=2.0)
+    plt.show()
+
+
+# Select four samples randomly for visualization.
+samples = list(json_dict.keys())
+n_samples = 4
```
Comment (fchollet) [inline examples/vision/keypoint_detection.py:216]: n_samples -> num_samples

File: examples/vision/keypoint_detection.py [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+import json, os
+
+"""
+## Define hyperparameters
+"""
+
+IMG_SIZE = 224
+BATCH_SIZE = 64
+EPOCHS = 5
+NB_KEYPOINTS = 24 * 2  # 24 pairs each having x and y coordinates
```
Comment (fchollet) [inline examples/vision/keypoint_detection.py:78]: NB_KEYPOINTS -> NUM_KEYPOINTS

File: examples/vision/keypoint_detection.py [hunk 9]
```diff
@@ -0,0 +1,440 @@
...
+            (new_image, new_kps_obj) = self.aug(image=current_image, keypoints=kps_obj)
+
+            # If the image is RGBA convert it to RGB.
+            if new_image.shape[-1] == 4:
+                new_image = Image.fromarray(new_image)
+                new_image = np.array(new_image.convert("RGB"))
+            batch_images[i,] = new_image
+
+            # Parse the coordinates from the new keypoint object.
+            kpTemp = []
```
Comment (fchollet) [inline examples/vision/keypoint_detection.py:300]: Use snake case

File: examples/vision/keypoint_detection.py [hunk 10]
```diff
@@ -0,0 +1,440 @@
...
+    ]
+)
+
+test_aug = iaa.Sequential([iaa.Resize(IMG_SIZE, interpolation="linear")])
+
+"""
+## Create training and validation splits
+"""
+
+train_keys, validation_keys = train_test_split(samples, test_size=0.15, random_state=42)
```
Comment (fchollet) [inline examples/vision/keypoint_detection.py:338]: Prefer doing this manually, we don't need to add a dependency on sklearn just for this

File: examples/vision/keypoint_detection.py [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+"""
+Predictions will likely improve with more training. 
+"""
+
+"""
+## Call for action
+
+* Try using other augmentation transforms from `imgaug` to investigate how that changes
+the results. 
+
```
Comment (fchollet) [inline examples/vision/keypoint_detection.py:435]: Remove line break

**Review by fchollet — COMMENTED** [review]
Interesting & useful example, thank you for the PR! 

Do you have a live Colab with the visualizations you get?

File: examples/vision/keypoint_detection.py (hunk 6 above)
Comment (sayakpaul) [reply examples/vision/keypoint_detection.py:108]: You can find it [here](https://colab.research.google.com/drive/1VY8CxDRW7PFw-ivseSx7bgoWpTnrQWOk?usp=sharing).

**Comment by sayakpaul** [comment]
@fchollet,

> Do you have a live Colab with the visualizations you get?

You can find it [here](https://colab.research.google.com/drive/1VY8CxDRW7PFw-ivseSx7bgoWpTnrQWOk?usp=sharing).

File: examples/vision/keypoint_detection.py [hunk 12]
```diff
@@ -0,0 +1,441 @@
...
+"""
+
+
+def get_model():
+    backbone = keras.applications.MobileNetV2(
+        weights="imagenet", include_top=False, input_shape=(IMG_SIZE, IMG_SIZE, 3)
+    )
+    inputs = layers.Input((IMG_SIZE, IMG_SIZE, 3))
+    x = keras.applications.mobilenet_v2.preprocess_input(inputs)
+    x = backbone(x, training=False)
```
Comment (srihari-humbarwadi) [inline examples/vision/keypoint_detection.py:377]: This would run the batch norm layers in inference mode, we think we should specify why this is being done

Comment (sayakpaul) [reply examples/vision/keypoint_detection.py:377]: I didn't specify it for brevity. I can add a sentence to refer to the fine-tuning guide where it's been explained.

Comment (fchollet) [reply examples/vision/keypoint_detection.py:377]: If the intent is to keep `backbone` frozen, we should do `backbone.trainable = False` (which will also run the BN layers in inference mode). An issue here is that we're adding randomly initialized layers on top of pretrained features, which largely negates the usefulness of the pretrained features.

File: examples/vision/keypoint_detection.py [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+    x = layers.Dropout(0.3)(x)
+    x = layers.SeparableConv2D(
+        NUM_KEYPOINTS, kernel_size=5, strides=1, activation="relu"
+    )(x)
+    outputs = layers.SeparableConv2D(
+        NUM_KEYPOINTS, kernel_size=3, strides=1, activation="sigmoid"
+    )(x)
+
+    return tf.keras.Model(inputs, outputs, name="keypoint_detector")
```
Comment (fchollet) [inline examples/vision/keypoint_detection.py:386]: Just `keras`, it's imported.

File: examples/vision/keypoint_detection.py (hunk 12 above)
Comment (sayakpaul) [reply examples/vision/keypoint_detection.py]: I went with `backbone.trainable = False` and added comments. 

> An issue here is that we're adding randomly initialized layers on top of pretrained features, which largely negates the usefulness of the pretrained features.

So, how would you suggest we handle this? In transfer learning settings, appending randomly initialized layer(s) on top of pre-trained layers is common.

File: examples/vision/keypoint_detection.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (8bitmp3) [reply examples/vision/keypoint_detection.py:2]: The Google Developer Docs Style Guide can be helpful https://developers.google.com/style/highlights

**Review by fchollet — COMMENTED** [review]
Thanks for the update! I added some copyedits. Besides this, everything looks good, I think we can add the generated files now.

**Comment by sayakpaul** [comment]
Sure. I will proceed accordingly. Thanks, @fchollet.

**Comment by sayakpaul** [comment]
@fchollet should be all set now.

**Review by fchollet — APPROVED** [review]
Thanks again for the great contribution! 👍

**Comment by sayakpaul** [comment]
Pleasure is all mine sir :)
